knoten/sensor_utils.py

```diff
@@ -0,0 +1,296 @@
+from knoten import csm, utils, bundle
+from csmapi import csmapi
+import numpy as np
+
+def phase_angle(image_pt, sensor):
+    """
+    Computes and returns phase angle, in degrees for a given image point.
+   
+    Phase Angle: The angle between the vector from the intersection point to
+    the observer (usually the spacecraft) and the vector from the intersection
+    point to the illuminator (usually the sun).
+
+    Parameters
+    ----------
+    image_pt : tuple
+               Pair of x, y (sample, line) coordinates in pixel space
+
+    sensor : object
+             A CSM compliant sensor model object
+
+    Returns
+    -------
+     : np.ndarray
+            phase angle in degrees
+    """
+    if not isinstance(image_pt, csmapi.ImageCoord):
+        image_pt = csmapi.ImageCoord(*image_pt)
+
+    sensor_state = csm.get_state(sensor, image_pt)
+    ground_pt = csm.generate_ground_point(0.0, image_pt, sensor)
+  
+    sunEcefVec = sensor.getIlluminationDirection(ground_pt)
```

Comment:
This is fairly dependent on shape model, we may want to allow users to pass a shapemodel in from the function signature. This is the case for any `generate_ground_point` call

Similarly with the illumination direction, for CSM we will have to track the sun/illunimator as it's own object so the information won't come from the sensor. We should likely allow someone to pass the illuminator in. This will be the case for any `getIlluminationDirection` call

Reply:
I was under the impression the CSM only supported ellipsoids therefore did not find the need for including a shapemodel. Let me know if that's incorrect. The implementation for using the `getIlluminationDirection` is from the CSMCamera::PhaseAngle function here: https://github.com/DOI-USGS/ISIS3/blob/21308692275a243ff58aeefce996cb3c0d8c357b/isis/src/base/objs/CSMCamera/CSMCamera.cpp#L869.

Reply:
Right, the CSM camera has it's own ellipsoid intersection routine against a biaxial ellipsoid. this is accessed through CSMs `imageToGround` routine. Something we will want to use at some point is a more detailed shapemodel to intersect. This is present in ISIS in the `CSMCamera` class [from line #141](https://github.com/DOI-USGS/ISIS3/blob/21308692275a243ff58aeefce996cb3c0d8c357b/isis/src/base/objs/CSMCamera/CSMCamera.cpp#L141) to line #171. There is some ISIS magic, but setting the look direction forces the shapemodel to compute a new intersection.

Since ISIS has a variety of shapemodels the intersection point can be different depending on the shapemodel. This sort of behavior is something we will eventually want in, what is growing to be, the knoten SET. 

As such a bare bones Ellipsoid shapemodel that exposes the vector normal from the shape, and an intersection function given a position and look vector would be useful as a first pass. Maybe not this sprint but at some point providing a distinct shapemodel will be very useful.

You're right about getIllunminationDirection. I think it works here, but we may want to provide another class for an Illuminator as the other functions that use it seem to need more information than just the illumination direction